caffe/src/caffe/layers/molgrid_data_layer.cpp

```diff
@@ -140,26 +140,45 @@ void MolGridDataLayer<Dtype>::getLigandChannels(int batch_idx, vector<short>& wh
 }
 
 template<typename Dtype>
-void MolGridDataLayer<Dtype>::getReceptorGradient(int batch_idx, vector<float3>& gradient, bool lrp)
+void MolGridDataLayer<Dtype>::getReceptorGradient(int batch_idx, vector<float3>& gradient)
 {
   gradient.resize(0);
   mol_info& mol = batch_transform[batch_idx].mol;
   for (unsigned i = 0, n = mol.atoms.size(); i < n; ++i)
     if (mol.whichGrid[i] < numReceptorTypes)
     {
-      if(lrp)
-      {
-          gradient.push_back(mol.gradient[i]);
-      }
-      else
-      {
-          gradient.push_back(-mol.gradient[i]);
-      }
+      gradient.push_back(mol.gradient[i]);
+    }
+}
+
+template<typename Dtype>
+void MolGridDataLayer<Dtype>::getMappedReceptorGradient(int batch_idx, unordered_map<string, float3>& gradient)
+{
+  mol_info& mol = batch_transform[batch_idx].mol;
+  for (unsigned i = 0, n = mol.atoms.size(); i < n; ++i)
+  {
+    std::stringstream ss;
+    ss << std::fixed << std::setprecision(3) << mol.atoms[i].x;
+    std::string rounded_x = ss.str();
+    ss.str("");
+    ss << std::fixed << std::setprecision(3) << mol.atoms[i].y;
+    std::string rounded_y = ss.str();
+    ss.str("");
+    ss << std::fixed << std::setprecision(3) << mol.atoms[i].z;
+    std::string rounded_z = ss.str();
+
+    string xyz = rounded_x + rounded_y + rounded_z;
+
+    if (mol.whichGrid[i] < numReceptorTypes)
```

Comment:
This check should happen _first_, otherwise you do a lot of unnecessary work in creating strings.